### PR TITLE
STM32: configure wake-up pins with GPIOs from devicetree for exiting Poweroff

### LINKED ISSUE #57323: RFC: introduce the STM32 system wakeup pins to exit low power modes
## Introduction
The stm32 mcu has several external wakeup pins (GPIO) that are able to trigger a wakeup from Standby or Shutdown low power modes.
The WakeUp pins are fixed GPIO pins. 
Each WakeUp pin is programmed previous to entering a low power standby/shutdown and will exit the system on edge-detection.
The stm32 can identify the source of WakeUp.

### Problem description
Depending on the stm32 serie, configuration of the WakeUp pins differs:
- the number of wakeup pin
- GPIO pin  position, including a multiplexer 4 -->1
- the trigger polarity : rising/falling edge
- pull-up/down 
- the configuration (control) and status of each wakeUp pin is made through the Power (PWR) instance of each stm32.

### Proposed change

1. Adding a new devicetree binding for each wakeUp pin (of each stm32 device)
2. Adding a new node for the PWR module of the stme32 device
3. Adding a new devicetree binding for using the WakeUp pins
4. Adding a software sequence to enable and configure each WakeUp pin in a new stm32_pwr module

## Detailed RFC

1. The list of possible wakeup pins depends on the stm32 device from 1 to 8. Each WakeUp pin is known by its gpio reference example (PC13). The stm32 associates (hardware fixed) this GPIO pin to a wakeup pin number : example PC13 is wakeup pin 2.
Each possible WakeUp pin of the stm32 is a node of the stm32 soc devicetree.
Because some wakeup pin configurations depend on the stm32 serie, several bindings are defined (edge rising/falling detection, pull-up down (stm32h7), multiplexing (stm32u5).
2. When using a Wakeup pin, it is listed  in a node of the devicetree of the target platform.
3. During system Initialisation, the pwr module is parsing this list of wakeUp pins and configures the stm32 PWR registers.

### Proposed change (Detailed)

- dts/bindings/power/st,stm32-wkup-pinX.yaml  for the wakeup pin node of the **stm32 soc** devicetree
For example on the stm32g4 serie, the system wakeup pin 2 is the GPIO PC13:

```
wakeup2_pc13: wakeup2_pc13 {
		compatible = "st,stm32-wkup-pin1";
		#wkpin-cells = <1>;
		syswakeup-pin = <2>;
};
```

- dts/bindings/power/st,stm32-pwr.yaml  for the pwr node of the **stm32 soc** devicetree
For example on the stm32g4 serie, the power node has 5 wakeup pins with configurable polarity

```
	pwr_wkup: pwr@40007000 {
		compatible = "st,stm32-pwr";
		reg = <0x40007000 0x400>; /* PWR register bank */
		clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>;
		pin-nb = <5>; /* 5 system WakeUp pins */
		pin-pol;  /* -->  the polarity of the pin for triggering is configurable */
		status = "disabled";
	};

```
- dts/bindings/power/st,stm32-pwr.yaml  for the list (_phandle-array_) of  wakeup pins of the pwr_wkup node in **stm32 board** device tree
For example on the nucleo stm32g474re : 
```
&pwr_wkup {
	status = "okay";
	wkpins = <&wakeup2_pc13 STM32_PWR_WAKEUP_RISING>;
};

```

- A new module is introduced /soc/arm/st_stm32/common/soc_pwr_config.c 
It configure and enable each wakeup pin listed in the `wkpins` property of the `&pwr_wkup` node.

### Dependencies
power management : CONFIG_PM
--> Using the system wakeup pins is pertinent when the the stm32 supports the standby (PM_STATE_STANDBY)  and/or shutdown (PM_STATE_SOFT_OFF) low power modes.

### Unresolved Questions

--> for each stm32 deice, the stm32XX.dtsi includes the complete list of wakeUp pins : that should be automatically generated in  the [hal_stm32](https://github.com/zephyrproject-rtos/hal_stm32)/[dts](https://github.com/zephyrproject-rtos/hal_stm32/tree/main/dts)/st/
--> Identify the wakeup source when exiting the low power (standby/shutdown) modes

### PR DESCRIPTION
**Pull request by ajarmouni-st**
This PR implements end-to-end configuration of GPIO pins as sources for STM32 PWR wake-up pins behind the scenes exclusively from devicetree information for all series using a newly introduced public `stm32_pwr_wkup_pin_cfg_gpio()` function inside `gpio_stm32.c` driver during initial GPIO pin configuration.

For this, we add a devicetree binding for the STM32 power contoller (PWR) peripheral which manages all device power supplies and power modes transitions, including the control of wake-up pins. We note that this initial PWR binding is focused on wake-up pins.

We also introduce a STM32 specific sample application that demonstrates how GPIO pins can be used to power on the system after a Poweroff, just like a reset pin.
The user only needs to add the `STM32_GPIO_WKUP` custom GPIO flag to the GPIO pin in "gpios" property of a DT device, such as a button or a sensor.
The sample powers off the system after a delay then the user can power it on again with the user button that has the `STM32_GPIO_WKUP` flag in DT.
Will only work on STM32 boards that support Zephyr POWEROFF (U5, WL, WB, & L4 SoC series), & also happen to have a user button whose GPIO pin is associated with a wake-up pin (see PWR node in SoC dtsi file).

For more information about how wake-up pins are managed & configured on a given STM32 SoC, please refer to the "Power control (PWR)" section in the Reference Manual. Information about GPIOs associated with wake-up pins can be found in the "pinouts and pin description" section in the datasheet.

### CODE REVIEW

**Comment by erwango** [comment]
@nandojve ^^

**Comment by ajarmouni-st** [comment]
Additional context: this PR introduced a STM32 custom GPIO flag, `ST_STM32_GPIO_WKUP`, to be used to signal to GPIO driver that a GPIO pin should be setup as a source for wake-up pins, instead of using the `wakeup-source` property, because using using a flag is more generic, & allows for use cases, beyond just GPIO Keys, where a DT device (ex: sensor) could have multiple GPIO pins in its `gpios`property, & only one of those GPIO pins is supposed to be used as wake-up source.

**Comment by ajarmouni-st** [comment]
Hi @bjarki-trackunit, I would appreciate your feedback, thanks!

File: samples/boards/stm32/power_mgmt/wkup_pins/src/main.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+int main(void)
+{
+	uint32_t cause;
+
+	hwinfo_get_reset_cause(&cause);
+	hwinfo_clear_reset_cause();
+
+	if (cause & RESET_BROWNOUT) {
+		printk("\nReset cause: wake-up pin or reset pin\n\n");
```
[the remark this answers is not in the document]
Comment (ajarmouni-st) [reply samples/boards/stm32/power_mgmt/wkup_pins/src/main.c:34]: Yes you are right, meant to use `RESET_PIN`. thanks!

**Comment by ajarmouni-st** [comment]
Rebase

File: samples/boards/stm32/power_mgmt/wkup_pins/src/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajarmouni-st) [reply samples/boards/stm32/power_mgmt/wkup_pins/src/main.c]: I removed reset cause completely cz it does not give the same result on all boards.

**Comment by ajarmouni-st** [comment]
Hi @henrikbrixandersen @fabiobaltieri could you please take a look? Thanks!

**Comment by erwango** [comment]
Moving assignee from @galak to @henrikbrixandersen as the owner of the next impacted subsystem (GPIO)

File: soc/st/stm32/common/stm32_wkup_pins.c [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiob)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioc)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiod)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioe)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiof)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiog)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioh)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioi)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioj)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiok)),
```
Comment (henrikbrixandersen) [inline soc/st/stm32/common/stm32_wkup_pins.c:171]: You could use macros for checking if the give nodelabel is enabled and save a few array entries on SoCs with fewer GPIO ports.

File: soc/st/stm32/common/stm32_wkup_pins.c
Comment (henrikbrixandersen) [inline soc/st/stm32/common/stm32_wkup_pins.c:1]: I'd probably put a few more comments into the implementations in this file.

**Review by henrikbrixandersen — DISMISSED** [review]

File: soc/st/stm32/common/stm32_wkup_pins.c
Comment (ajarmouni-st) [reply soc/st/stm32/common/stm32_wkup_pins.c:1]: Will do, thanks for reviewing !

File: samples/boards/stm32/power_mgmt/wkup_pins/src/main.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/poweroff.h>
+
+#define WAIT_TIME_US (1 << 22)
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/wkup_pins/src/main.c:14]: Please provide a number easier to read, or at least add a comment.

File: include/zephyr/dt-bindings/gpio/st-stm32-gpio.h
Comment (erwango) [inline include/zephyr/dt-bindings/gpio/st-stm32-gpio.h:1]: Please  rename to stm32-gpio.h

File: include/zephyr/dt-bindings/gpio/st-stm32-gpio.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * @ingroup gpio_interface
+ * @{
+ */
+
+/**
+ * Configures a GPIO pin to power on the system after Poweroff.
+ * This flag is reserved to GPIO pins that are associated with wake-up pins
+ * in STM32 PWR devicetree node, through the property "wkup-gpios".
+ */
+#define ST_STM32_GPIO_WKUP         (1 << 8)
```
Comment (erwango) [inline include/zephyr/dt-bindings/gpio/st-stm32-gpio.h:27]: Please rename to STM32_GPIO... to stick with existing convention on flags naming

**Review by erwango — CHANGES_REQUESTED** [review]
Otherwise LGTM in the general lines.

**Comment by zafersn** [comment]
> Hi @zafersn, was it a `nucleo_u575zi_q` board? I just tested it on my side & the user button is working.
> 
> ```
> / {
> 	aliases {
> 		wkup-src = &user_button;
> 	};
> };
> 
> &user_button {
> 	gpios = <&gpioc 13 (GPIO_ACTIVE_HIGH | ST_STM32_GPIO_WKUP)>;
> };
> 
> &pwr {
> 	status = "okay";
> };
> ```
> 
> Just to verify quickly that the system actually goes into shutdown mode, LED 1 will be turned off.
> 
> Will try to verify current consumption on `nucleo_u575zi_q`.

Thanks for quick reply @ajarmouni-st : 

I see the differences now that you have` (GPIO_ACTIVE_HIGH | ST_STM32_GPIO_WKUP)` flag configs. I tried with this configs and now button works. and also I have new current consumtion of 5.5uA:

image

**Comment by ajarmouni-st** [comment]
> also I have new current consumtion of 5.5uA but it is still not even close to 160nA shutdown mode consumption.

@zafersn I got 1 uA but I don't know if it's the real consu. or it's bcz my ammeter cannot go lower.

File: soc/st/stm32/common/stm32_wkup_pins.c (hunk 2 above)
Comment (ajarmouni-st) [reply soc/st/stm32/common/stm32_wkup_pins.c:171]: In addition to getting the device ptr of each GPIO port, this table is also used to get the index of a given GPIO port, so keeping the NULL entries is desired.

File: soc/st/stm32/common/stm32_wkup_pins.c [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+#define WKUP_PIN_CFG_DT_COMMA(wkup_pin_id) WKUP_PIN_CFG_DT(wkup_pin_id),
+
+#if PWR_STM32_WKUP_PINS_PUPD_CFG
+
+#define STM32_GPIOS_NB               16
+
+#if defined(CONFIG_SOC_SERIES_STM32U5X) || defined(CONFIG_SOC_SERIES_STM32WBAX)
+#define PWR_STM32_GPIO_BIT_DEF(i, _) LL_PWR_GPIO_PIN_##i
+#else
+#define PWR_STM32_GPIO_BIT_DEF(i, _) LL_PWR_GPIO_BIT_##i
```
Comment (erwango) [inline soc/st/stm32/common/stm32_wkup_pins.c:121]: Could be (1<<i)

File: include/zephyr/dt-bindings/gpio/st-stm32-gpio.h
Comment (erwango) [reply include/zephyr/dt-bindings/gpio/st-stm32-gpio.h:1]: This is not all and no stm32 at least.

**Comment by zafersn** [comment]
> > also I have new current consumtion of 5.5uA but it is still not even close to 160nA shutdown mode consumption.
> 
> @zafersn I got 1 uA but I don't know if it's the real consu. or it's bcz my ammeter cannot go lower.
> 
> In any case, feel free to open an issue regarding current consumption in shutdown mode on STM32U575 & we can continue the discussion there.

Ok thanks for this information.

BTW, I also tried to get reset cause information.

`hwinfo_get_reset_cause(&cause);`

cause return with 0 after waking up from LL_PWR_SHUTDOWN_MODE
Is there any possibility to get the correct cause info?

**Comment by ajarmouni-st** [comment]
> BTW, I also tried to get reset cause information.
> 
> `hwinfo_get_reset_cause(&cause);`
> 
> cause return with 0 after waking up from LL_PWR_SHUTDOWN_MODE Is there any possibility to get the correct cause info?

@zafersn I had it in the PR previously but decided to remove it cz I couldn't get consistent results on all boards, see https://github.com/zephyrproject-rtos/zephyr/compare/4235e22f6c855f3c58386a04812f57ce084618a0..601e1ded1dc99273ab9ce6c4b253fdfd4363279a

**Review by erwango — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**